### PR TITLE
Fix small_dt increments error

### PR DESCRIPTION
**Pull request by CKehl**

### CODE REVIEW

**Comment by CKehl** [comment]
Still working on it - I have a new barebone of the main function taht is resilient to floating-point (in)accuracy and limited-precision numer comparisons. Still need to iron-out corner cases. I push the most recent version up, but it's prototypical. It still failts when attempting to delete a particle from within the kernel.

**Comment by CKehl** [comment]
remaining issue/bug: `test_pset_repeated_release_delayed_adding_deleting` in **tests/test_particle_file.py**:
```
_______________________________________________________________________________________ test_pset_repeated_release_delayed_adding_deleting[10--1-2-scipy-timearr] _______________________________________________________________________________________

>   ???
E   ValueError: cannot reshape array of size 2 into shape (10,)

netCDF4/_netCDF4.pyx:4866: ValueError

During handling of the above exception, another exception occurred:

type = 'timearr', fieldset = <parcels.fieldset.FieldSet object at 0x7fec726ee210>, mode = 'scipy', repeatdt = 2, tmpdir = local('/tmp/pytest-of-ckehl/pytest-6/test_pset_repeated_release_del37'), dt = -1, maxvar = 10, runtime = 10

    @pytest.mark.parametrize('type', ['repeatdt', 'timearr'])
    @pytest.mark.parametrize('mode', ['scipy', 'jit'])
    @pytest.mark.parametrize('repeatdt', range(1, 3))
    @pytest.mark.parametrize('dt', [-1, 1])
    @pytest.mark.parametrize('maxvar', [2, 4, 10])
    def test_pset_repeated_release_delayed_adding_deleting(type, fieldset, mode, repeatdt, tmpdir, dt, maxvar, runtime=10):
        fieldset.maxvar = maxvar
    
        class MyParticle(ptype[mode]):
            sample_var = Variable('sample_var', initial=0.)
            v_once = Variable('v_once', dtype=np.float64, initial=0., to_write='once')
        if type == 'repeatdt':
            pset = ParticleSet(fieldset, lon=[0], lat=[0], pclass=MyParticle, repeatdt=repeatdt)
        elif type == 'timearr':
            pset = ParticleSet(fieldset, lon=np.zeros(runtime), lat=np.zeros(runtime), pclass=MyParticle, time=list(range(runtime)))
        outfilepath = tmpdir.join("pfile_repeated_release.nc")
        pfile = pset.ParticleFile(outfilepath, outputdt=abs(dt))
    
        def IncrLon(particle, fieldset, time):
            particle.sample_var += 1.
            if particle.sample_var > fieldset.maxvar:
                particle.delete()
        for i in range(runtime):
            pset.execute(IncrLon, dt=dt, runtime=1, output_file=pfile)
    
>       ncfile = close_and_compare_netcdffiles(outfilepath, pfile)

tests/test_particle_file.py:183: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
tests/test_particle_file.py:31: in close_and_compare_netcdffiles
    convert.convert_npydir_to_netcdf(ofile.tempwritedir_base)
/var/scratch/test_env/lib/python3.7/site-packages/parcels-2.1.5.dev48+gfeef2c8.d20200309-py3.7.egg/parcels/scripts/convert_npydir_to_netcdf.py:26: in convert_npydir_to_netcdf
    pfile.close(delete_tempfiles)
/var/scratch/test_env/lib/python3.7/site-packages/parcels-2.1.5.dev48+gfeef2c8.d20200309-py3.7.egg/parcels/particlefile.py:200: in close
    self.export()
/var/scratch/test_env/lib/python3.7/site-packages/parcels-2.1.5.dev48+gfeef2c8.d20200309-py3.7.egg/parcels/particlefile.py:382: in export
    getattr(self, var)[:] = self.read_from_npy(global_file_list_once, 1, var)
netCDF4/_netCDF4.pyx:4868: in netCDF4._netCDF4.Variable.__setitem__
    ???
<__array_function__ internals>:6: in broadcast_to
    ???
/var/scratch/bin-x86_64/anaconda/envs/OpenParcels_MPI/lib/python3.7/site-packages/numpy/lib/stride_tricks.py:182: in broadcast_to
    return _broadcast_to(array, shape, subok=subok, readonly=True)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

array = array([[0.],
       [0.]]), shape = (10,), subok = False, readonly = True

    def _broadcast_to(array, shape, subok, readonly):
        shape = tuple(shape) if np.iterable(shape) else (shape,)
        array = np.array(array, copy=False, subok=subok)
        if not shape and array.shape:
            raise ValueError('cannot broadcast a non-scalar to a scalar array')
        if any(size < 0 for size in shape):
            raise ValueError('all elements of broadcast shape must be non-'
                             'negative')
        extras = []
        it = np.nditer(
            (array,), flags=['multi_index', 'refs_ok', 'zerosize_ok'] + extras,
>           op_flags=['readonly'], itershape=shape, order='C')
E       ValueError: input operand has more dimensions than allowed by the axis remapping

/var/scratch/bin-x86_64/anaconda/envs/OpenParcels_MPI/lib/python3.7/site-packages/numpy/lib/stride_tricks.py:127: ValueError
----------------------------------------------------------------------------------------------------------------- Captured stdout call ------------------------------------------------------------------------------------------------------------------
p.time: 0.0; endtime: 8.0, dt: -1, p.dt: -1
p.time: 1.0; endtime: 8.0, dt: -1, p.dt: -1
p.time: 2.0; endtime: 8.0, dt: -1, p.dt: -1
p.time: 3.0; endtime: 8.0, dt: -1, p.dt: -1
p.time: 4.0; endtime: 8.0, dt: -1, p.dt: -1
p.time: 5.0; endtime: 8.0, dt: -1, p.dt: -1
p.time: 6.0; endtime: 8.0, dt: -1, p.dt: -1
p.time: 7.0; endtime: 8.0, dt: -1, p.dt: -1
p.time: 8.0; endtime: 8.0, dt: -1, p.dt: -1
p.time: 9.0; endtime: 8.0, dt: -1, p.dt: -1
p.time: 8.0; endtime: 8.0, dt: -1, p.dt: -1
p.time: 8.0; endtime: 8.0, dt: -1, p.dt: -1
p.time: 8.0; endtime: 7.0, dt: -1, p.dt: -1
p.time: 8.0; endtime: 7.0, dt: -1, p.dt: -1
p.time: 7.0; endtime: 6.0, dt: -1, p.dt: -1
p.time: 7.0; endtime: 6.0, dt: -1, p.dt: -1
p.time: 6.0; endtime: 5.0, dt: -1, p.dt: -1
p.time: 6.0; endtime: 5.0, dt: -1, p.dt: -1
p.time: 5.0; endtime: 4.0, dt: -1, p.dt: -1
p.time: 5.0; endtime: 4.0, dt: -1, p.dt: -1
p.time: 4.0; endtime: 3.0, dt: -1, p.dt: -1
p.time: 4.0; endtime: 3.0, dt: -1, p.dt: -1
p.time: 3.0; endtime: 2.0, dt: -1, p.dt: -1
p.time: 3.0; endtime: 2.0, dt: -1, p.dt: -1
p.time: 2.0; endtime: 1.0, dt: -1, p.dt: -1
p.time: 2.0; endtime: 1.0, dt: -1, p.dt: -1
p.time: 1.0; endtime: 0.0, dt: -1, p.dt: -1
p.time: 1.0; endtime: 0.0, dt: -1, p.dt: -1
p.time: 0.0; endtime: -1.0, dt: -1, p.dt: -1
p.time: 0.0; endtime: -1.0, dt: -1, p.dt: -1
```

according to my understanding, the `endtime` should (in the non-failure case) have the same sign as `dt`. I don't know why that is not the case here.

File: parcels/codegenerator.py [hunk 1]
```diff
@@ -922,48 +922,81 @@ def generate(self, funcname, field_args, const_args, kernel_ast, c_include):
...
-        dt_pos = c.Assign("__dt", "fmin(fabs(particles[p].dt), fabs(endtime - particles[p].time))")
+        sign_end_part = c.Assign("sign_end_part", "(endtime - particles[p].time) > 0 ? 1 : -1")
+        reset_res_state = c.Assign("res", "particles[p].state")
+        update_state = c.Assign("particles[p].state", "res")
+        update_pdt = c.If("_next_dt_set == 1",
+                          c.Block([c.Assign("_next_dt_set", "0"), c.Assign("particles[p].dt", "_next_dt")]))
+
+        dt_pos = c.Assign("__dt", "fmin(fabs(particles[p].dt), fabs(endtime - particles[p].time))")                   # original
+        # dt_pos = c.Assign("__dt", "fmin(fabs(particles[p].dt), (endtime - particles[p].time) * sign_dt)")           # reint
+        # dt_pos = c.Assign("__dt", "fmax(fmin(fabs(particles[p].dt), (endtime - particles[p].time) * sign_dt), 0)")  # christian
```
Comment (erikvansebille) [inline parcels/codegenerator.py]: Remove these commented lines?

File: parcels/codegenerator.py [hunk 2]
```diff
@@ -922,48 +922,81 @@ def generate(self, funcname, field_args, const_args, kernel_ast, c_include):
...
+
         pdt_eq_dt_pos = c.Assign("__pdt_prekernels", "__dt * sign_dt")
         partdt = c.Assign("particles[p].dt", "__pdt_prekernels")
-        dt_0_break = c.If("particles[p].dt == 0", c.Statement("break"))
-        notstarted_continue = c.If("(sign_end_part != sign_dt) && (particles[p].dt != 0)",
-                                   c.Statement("continue"))
+        check_pdt = c.If("(res == SUCCESS) & !is_equal_dbl(__pdt_prekernels, particles[p].dt)", c.Assign("res", "REPEAT"))
+
+        dt_0_break = c.If("is_zero_dbl(particles[p].dt)", c.Statement("break"))
+        # dt_0_break = c.If("is_zero_dbl(dt)", c.Statement("break"))
```
Comment (erikvansebille) [inline parcels/codegenerator.py]: Remove this commented line?

File: parcels/particlefile.py [hunk 3]
```diff
@@ -234,6 +234,11 @@ def convert_pset_to_dict(self, pset, time, deleted_only=False):
             else:
                 if deleted_only:
                     pset_towrite = pset
+                # == commented due to git rebase to master 27 02 2020 == #
+                # elif pset[0].dt > 0:
+                #     pset_towrite = [p for p in pset if time - p.dt/2 <= p.time < time + p.dt and np.isfinite(p.id)]
+                # else:
+                #     pset_towrite = [p for p in pset if time + p.dt < p.time <= time - p.dt/2 and np.isfinite(p.id)]
```
Comment (erikvansebille) [inline parcels/particlefile.py]: Can this commented code be removed?

File: parcels/particleset.py [hunk 4]
```diff
@@ -56,6 +56,7 @@ def __init__(self, fieldset, pclass=JITParticle, lon=None, lat=None, depth=None,
         self.fieldset = fieldset
         self.fieldset.check_complete()
         partitions = kwargs.pop('partitions', None)
+        self.aborted = False  # TODO is this necessary?
```
Comment (erikvansebille) [inline parcels/particleset.py]: Check this `#TODO` message? Is it indeed necessary?

File: parcels/particleset.py [hunk 5]
```diff
@@ -520,7 +521,7 @@ def execute(self, pyfunc=AdvectionRK4, endtime=None, runtime=None, dt=1.,
             walltime_start = time_module.time()
         if verbose_progress:
             pbar = self.__create_progressbar(_starttime, endtime)
-        while (time < endtime and dt > 0) or (time > endtime and dt < 0) or dt == 0:
+        while ((time < endtime and dt > 0) or (time > endtime and dt < 0)) and not self.aborted or dt == 0:
```
Comment (erikvansebille) [inline parcels/particleset.py]: So this `self.aborted` is only used in the unit test? Do we then really need it?

File: tests/test_advection.py [hunk 6]
```diff
@@ -134,12 +134,13 @@ def SubmergeParticle(particle, fieldset, time):
         particle.depth = 0
         AdvectionRK4(particle, fieldset, time)  # perform a 2D advection because vertical flow will always push up in this case
         particle.time = time + particle.dt  # to not trigger kernels again, otherwise infinite loop
+        particle.succeeded()
```
Comment (erikvansebille) [inline tests/test_advection.py]: This `self.succeeded` is only used in this unit tests? Why not directly assign ?
```suggestion
        particle.status = ErrorCode.Success
```

File: tests/test_advection.py [hunk 7]
```diff
@@ -134,12 +134,13 @@ def SubmergeParticle(particle, fieldset, time):
...
 
     recovery_dict = {ErrorCode.ErrorOutOfBounds: DeleteParticle}
     if wErrorThroughSurface:
         recovery_dict[ErrorCode.ErrorThroughSurface] = SubmergeParticle
 
-    pset = ParticleSet(fieldset=fieldset, pclass=ptype[mode], lon=0.5, lat=0.5, depth=0.9)
+    pset = ParticleSet(fieldset=fieldset, pclass=ptype[mode], lon=[0.5, ], lat=[0.5, ], depth=[0.9, ])
```
Comment (erikvansebille) [inline tests/test_advection.py]: Why is this change necessary? It should be possible to define `lon`, `lat` etc as scalars?

File: tests/test_particle_file.py [hunk 8]
```diff
@@ -181,8 +181,9 @@ def IncrLon(particle, fieldset, time):
         particle.sample_var += 1.
         if particle.sample_var > fieldset.maxvar:
             particle.delete()
+
     for i in range(runtime):
-        pset.execute(IncrLon, dt=dt, runtime=1., output_file=pfile)
+        pset.execute(pset.Kernel(IncrLon, delete_cfiles=False), dt=dt, runtime=1., output_file=pfile)
```
Comment (erikvansebille) [inline tests/test_particle_file.py]: This change is not necessary anymore? Was only there for debugging purposes (keeping the cfiles)?

File: tests/test_smalldt.py [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+
+Created on Thu Feb 27 14:44:03 2020
+
+@author: reint fischer
+"""
+
+# Local Parcels path for Reint
+# import sys
+#
+# sys.path.insert(0, "\\Users\\Gebruiker\\Documents\\GitHub\\parcels\\")  # Set path to find the newest parcels code
```
Comment (erikvansebille) [inline tests/test_smalldt.py]: This can be removed here?

File: tests/test_smalldt.py [hunk 10]
```diff
@@ -0,0 +1,145 @@
+# -*- coding: utf-8 -*-
```
Comment (erikvansebille) [inline tests/test_smalldt.py]: This is a fairly large/wieldy script for a unit test. It's also different in style from most other unit tests. Would it be an idea to simplify and put in one of the other unit test files?

File: parcels/codegenerator.py (hunk 1 above)
Comment (CKehl) [reply parcels/codegenerator.py]: yes, I can do that - left it in until getting the final OK from Reint

File: parcels/codegenerator.py (hunk 2 above)
Comment (CKehl) [reply parcels/codegenerator.py]: yes

File: parcels/particlefile.py (hunk 3 above)
Comment (CKehl) [reply parcels/particlefile.py]: On that one, I am unsure. it"s in there since 3 weeks in the master cause I wanted to first test the patch with our our other applications. As we still have little feedback on the working of recent changes, I'd still rather keep it in until we can really confirm that the patch here works.

We can remove it, but I wouldn't do it right now but rather in a few weeks.

File: parcels/particleset.py (hunk 4 above)
Comment (CKehl) [reply parcels/particleset.py]: the variable IS needed, the TODO message not. I may need to go over Reint's smaller changes still ...

File: tests/test_advection.py (hunk 6 above)
Comment (CKehl) [reply tests/test_advection.py]: well, ideally setting the status is done via function call and NOT via direct setting. Has to to with that variable being marked as 'read_only' in particle.py. What's your take on that ? I think it's actually a good idea to only modify the status internally and handle it just like a private member variable. Also, that (i.e. modifying the state with some code) is really not something the user shall be doing ...

File: tests/test_advection.py (hunk 7 above)
Comment (CKehl) [reply tests/test_advection.py]: actually, yes, that was when I tested the function to find the error, but it can be restored to its original state.

File: tests/test_particle_file.py (hunk 8 above)
Comment (CKehl) [reply tests/test_particle_file.py]: agreed, thanks for the feedback, need to rewind that change

File: tests/test_smalldt.py (hunk 9 above)
Comment (CKehl) [reply tests/test_smalldt.py]: yes, I'll do that

File: tests/test_smalldt.py (hunk 10 above)
Comment (CKehl) [reply tests/test_smalldt.py]: really ? I think it is a good, short and quick script. Also, it does replicate the error of small dt incrementals in a way that can be tested. Still, if you have a better idea I'm all ears.

File: parcels/particlefile.py (hunk 3 above)
Comment (erikvansebille) [reply parcels/particlefile.py]: This will be changed anyway when we merge #678
